unstable test test_search_beam_position.test_search_multiple

Sometimes it passes, sometimes it:
```
run_in_tmpdir = local('/scratch/jenkins_agent/workspace/dials_tests/tests/pytest/popen-gw2/test_search_multiple0')
dials_regression = '/dls/science/groups/scisoft/DIALS/repositories/git-reference/dials_regression'

    def test_search_multiple(run_in_tmpdir, dials_regression):
        """Perform a beam-centre search and check that the output is sane.
    
        Do the following:
        1. Run dials.search_beam_centre on two datablocks and two pickled
        reflection tables, as output by dials.find_spots;
          a) Check that the program exits correctly;
          b) Check that it produces the expected output datablock.
        2. Check that the beam centre search has resulted in the expected shift
        in detector origin.
        """
    
        data_dir = os.path.join(dials_regression, "indexing_test_data", "trypsin")
        pickle_path1 = os.path.join(data_dir, "strong_P1_X6_1_0-1.pickle")
        pickle_path2 = os.path.join(data_dir, "strong_P1_X6_2_0-1.pickle")
        experiments_path1 = os.path.join(data_dir, "datablock_P1_X6_1.json")
        experiments_path2 = os.path.join(data_dir, "datablock_P1_X6_2.json")
    
        args = [experiments_path1, experiments_path2, pickle_path1, pickle_path2]
        search_beam_position.run(args)
        assert os.path.exists("optimised.expt")
    
        experiments = load.experiment_list(experiments_path1, check_format=False)
        optimised_experiments = load.experiment_list("optimised.expt", check_format=False)
        detector_1 = experiments[0].detector
        detector_2 = optimised_experiments[0].detector
        shift = scitbx.matrix.col(detector_1[0].get_origin()) - scitbx.matrix.col(
            detector_2[0].get_origin()
        )
>       assert shift.elems == pytest.approx((0.037, 0.061, 0.0), abs=1e-1)
E       assert (3.7479447822...48082165, 0.0) == approx((0.037 ...0 +- 1.0e-01))
E         -(3.7479447822514658, 2.964508548082165, 0.0)
E         +approx((0.037 +- 1.0e-01, 0.061 +- 1.0e-01, 0.0 +- 1.0e-01))

../../../../build_dials/modules/dials/test/command_line/test_search_beam_position.py:86: AssertionError
```

--- a/command_line/search_beam_position.py
+++ b/command_line/search_beam_position.py
@@ -421,7 +421,8 @@ def discover_better_experimental_model(
             futures.append(pool.submit(run_dps, expt, refl, max_cell))
         solution_lists = []
         amax_list = []
-        for future in concurrent.futures.as_completed(futures):
+        concurrent.futures.wait(futures, return_when=concurrent.futures.ALL_COMPLETED)
+        for future in futures:
             result = future.result()
             if result.get("solutions"):
                 solution_lists.append(result["solutions"])

--- a/command_line/search_beam_position.py
+++ b/command_line/search_beam_position.py
@@ -3,6 +3,7 @@ from __future__ import absolute_import, division, print_function
 import cmath
 import concurrent.futures
 import copy
+import itertools
 import logging
 import math
 import random
@@ -354,7 +355,7 @@ def run_dps(experiment, spots_mm, max_cell):
     # There must be at least 3 solutions to make a set, otherwise return empty result
     if len(solutions) < 3:
         return {}
-    return dict(solutions=flex.vec3_double([s.dvec for s in solutions]), amax=DPS.amax)
+    return dict(solutions=flex.vec3_double(s.dvec for s in solutions), amax=DPS.amax)
 
 
 def discover_better_experimental_model(
@@ -416,14 +417,11 @@ def discover_better_experimental_model(
         max_cell = params.max_cell
 
     with concurrent.futures.ProcessPoolExecutor(max_workers=nproc) as pool:
-        futures = []
-        for expt, refl in zip(experiments, refl_lists):
-            futures.append(pool.submit(run_dps, expt, refl, max_cell))
         solution_lists = []
         amax_list = []
-        concurrent.futures.wait(futures, return_when=concurrent.futures.ALL_COMPLETED)
-        for future in futures:
-            result = future.result()
+        for result in pool.map(
+            run_dps, experiments, refl_lists, itertools.repeat(max_cell)
+        ):
             if result.get("solutions"):
                 solution_lists.append(result["solutions"])
                 amax_list.append(result["amax"])

--- a/command_line/search_beam_position.py
+++ b/command_line/search_beam_position.py
@@ -117,20 +117,25 @@ def optimize_origin_offset_local_scope(
         plot_px_sz *= wide_search_binning
         grid = max(1, int(mm_search_scope / plot_px_sz))
         widegrid = 2 * grid + 1
-        scores = flex.double()
-        for y in range(-grid, grid + 1):
-            for x in range(-grid, grid + 1):
-                new_origin_offset = x * plot_px_sz * beamr1 + y * plot_px_sz * beamr2
-                score = 0
-                for i in range(len(experiments)):
-                    score += _get_origin_offset_score(
-                        new_origin_offset,
-                        solution_lists[i],
-                        amax_lists[i],
-                        reflection_lists[i],
-                        experiments[i],
-                    )
-                scores.append(score)
+
+        def get_experiment_score_for_coord(x, y):
+            new_origin_offset = x * plot_px_sz * beamr1 + y * plot_px_sz * beamr2
+            return sum(
+                _get_origin_offset_score(
+                    new_origin_offset,
+                    solution_lists[i],
+                    amax_lists[i],
+                    reflection_lists[i],
+                    experiment,
+                )
+                for i, experiment in enumerate(experiments)
+            )
+
+        scores = flex.double(
+            get_experiment_score_for_coord(x, y)
+            for y in range(-grid, grid + 1)
+            for x in range(-grid, grid + 1)
+        )
 
         def igrid(x):
             return x - (widegrid // 2)
@@ -269,8 +274,6 @@ def _get_origin_offset_score(
     # positions not to the correct RS position => reset any fixed rotation
     # to identity - copy in case called from elsewhere
 
-    gonio = copy.deepcopy(experiment.goniometer)
-    gonio.set_fixed_rotation((1, 0, 0, 0, 1, 0, 0, 0, 1))
     spots_mm.map_centroids_to_reciprocal_space([experiment])
     return _sum_score_detail(spots_mm["rlp"], solutions, amax=amax)
 

--- a/command_line/search_beam_position.py
+++ b/command_line/search_beam_position.py
@@ -153,43 +153,43 @@ def optimize_origin_offset_local_scope(
     else:
         wide_search_offset = None
 
-    # DO A SIMPLEX MINIMIZATION
-
-    class test_simplex_method(object):
-        def __init__(selfOO, wide_search_offset=None):
-            selfOO.starting_simplex = []
-            selfOO.n = 2
-            selfOO.wide_search_offset = wide_search_offset
-            for ii in range(selfOO.n + 1):
-                selfOO.starting_simplex.append(flex.random_double(selfOO.n))
-            selfOO.optimizer = simplex_opt(
-                dimension=selfOO.n,
-                matrix=selfOO.starting_simplex,
-                evaluator=selfOO,
+    # Do a simplex minimization
+    class simplex_minimizer(object):
+        def __init__(self, wide_search_offset):
+            self.n = 2
+            self.wide_search_offset = wide_search_offset
+            self.starting_simplex = [  # guaranteed to be random
+                flex.double([0.11505456638977896, 0.6090665392794814]),
+                flex.double([0.13339096418598828, 0.24058961996534878]),
+                flex.double([0.3271390558111398, 0.8591374909485977]),
+            ]
+            self.optimizer = simplex_opt(
+                dimension=self.n,
+                matrix=self.starting_simplex,
+                evaluator=self,
                 tolerance=1e-7,
             )
-            selfOO.x = selfOO.optimizer.get_solution()
-            selfOO.offset = selfOO.x[0] * 0.2 * beamr1 + selfOO.x[1] * 0.2 * beamr2
-            if selfOO.wide_search_offset is not None:
-                selfOO.offset += selfOO.wide_search_offset
+            self.x = self.optimizer.get_solution()
+            self.offset = self.x[0] * 0.2 * beamr1 + self.x[1] * 0.2 * beamr2
+            if self.wide_search_offset is not None:
+                self.offset += self.wide_search_offset
 
-        def target(selfOO, vector):
+        def target(self, vector):
             trial_origin_offset = vector[0] * 0.2 * beamr1 + vector[1] * 0.2 * beamr2
-            if selfOO.wide_search_offset is not None:
-                trial_origin_offset += selfOO.wide_search_offset
+            if self.wide_search_offset is not None:
+                trial_origin_offset += self.wide_search_offset
             target = 0
-            for i in range(len(experiments)):
+            for i, experiment in enumerate(experiments):
                 target -= _get_origin_offset_score(
                     trial_origin_offset,
                     solution_lists[i],
                     amax_lists[i],
                     reflection_lists[i],
-                    experiments[i],
+                    experiment,
                 )
             return target
 
-    MIN = test_simplex_method(wide_search_offset=wide_search_offset)
-    new_offset = MIN.offset
+    new_offset = simplex_minimizer(wide_search_offset).offset
 
     if plot_search_scope:
         plot_px_sz = experiments[0].get_detector()[0].get_pixel_size()[0]
@@ -199,13 +199,13 @@ def optimize_origin_offset_local_scope(
             for x in range(-grid, grid + 1):
                 new_origin_offset = x * plot_px_sz * beamr1 + y * plot_px_sz * beamr2
                 score = 0
-                for i in range(len(experiments)):
+                for i, experiment in enumerate(experiments):
                     score += _get_origin_offset_score(
                         new_origin_offset,
                         solution_lists[i],
                         amax_lists[i],
                         reflection_lists[i],
-                        experiments[i],
+                        experiment,
                     )
                 scores.append(score)
 

--- a/command_line/search_beam_position.py
+++ b/command_line/search_beam_position.py
@@ -164,14 +164,9 @@ def optimize_origin_offset_local_scope(
         def __init__(self, wide_search_offset):
             self.n = 2
             self.wide_search_offset = wide_search_offset
-            self.starting_simplex = [  # guaranteed to be random
-                flex.double([0.11505456638977896, 0.6090665392794814]),
-                flex.double([0.13339096418598828, 0.24058961996534878]),
-                flex.double([0.3271390558111398, 0.8591374909485977]),
-            ]
             self.optimizer = simplex_opt(
                 dimension=self.n,
-                matrix=self.starting_simplex,
+                matrix=[flex.random_double(self.n) for _ in range(self.n + 1)],
                 evaluator=self,
                 tolerance=1e-7,
             )

